crates/provider/src/native.rs

```diff
@@ -1,833 +1,1522 @@
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct NativeNamespace<FS: FileSystem + Send + Sync + Clone> {
+    id: String,
+    base_dir: PathBuf,
+    inner: Arc<RwLock<NativeNamespaceInner<FS>>>,
+    filesystem: FS,
+    provider: WeakNativeProvider<FS>,
+}
+
+#[derive(Debug)]
+struct NativeNamespaceInner<FS: FileSystem + Send + Sync + Clone> {
+    nodes: HashMap<String, NativeNode<FS>>,
+}
+
+#[derive(Debug, Clone)]
+struct WeakNativeNamespace<FS: FileSystem + Send + Sync + Clone> {
+    inner: Weak<RwLock<NativeNamespaceInner<FS>>>,
+}
+
+#[async_trait]
+impl<FS: FileSystem + Send + Sync + Clone + 'static> ProviderNamespace for NativeNamespace<FS> {
+    fn id(&self) -> &str {
+        &self.id
     }
 
-    async fn get_node_info(&mut self, pod_name: String) -> Result<(IpAddr, Port), ProviderError> {
-        let host_port = self.get_port_mapping(P2P_PORT, pod_name).await?;
-        Ok((LOCALHOST, host_port))
+    fn base_dir(&self) -> &PathBuf {
+        &self.base_dir
     }
 
-    async fn get_node_ip(&self) -> Result<IpAddr, ProviderError> {
-        Ok(LOCALHOST)
+    async fn nodes(&self) -> HashMap<String, DynNode> {
+        self.inner
+            .read()
+            .await
+            .nodes
+            .clone()
+            .into_iter()
+            .map(|(id, node)| (id, Arc::new(node) as DynNode))
+            .collect()
     }
 
-    async fn run_command(
-        &self,
-        mut args: Vec<String>,
-        opts: NativeRunCommandOptions,
-    ) -> Result<RunCommandResponse, ProviderError> {
-        if let Some(arg) = args.get(0) {
-            if arg == "bash" {
-                args.remove(0);
-            }
+    async fn spawn_node(&self, options: SpawnNodeOptions) -> Result<DynNode, ProviderError> {
+        let mut inner = self.inner.write().await;
+
+        if inner.nodes.contains_key(&options.name) {
+            return Err(ProviderError::DuplicatedNodeName(options.name));
         }
 
-        // -c is already used in the process::Command to execute the command thus
-        // needs to be removed in case provided
-        if let Some(arg) = args.get(0) {
-            if arg == "-c" {
-                args.remove(0);
-            }
+        // create node directories and filepaths
+        let base_dir_raw = format!("{}/{}", &self.base_dir.to_string_lossy(), &options.name);
+        let base_dir = PathBuf::from(&base_dir_raw);
+        let log_path = PathBuf::from(format!("{}/{}.log", base_dir_raw, &options.name));
+        let config_dir = PathBuf::from(format!("{}{}", base_dir_raw, NODE_CONFIG_DIR));
+        let data_dir = PathBuf::from(format!("{}{}", base_dir_raw, NODE_DATA_DIR));
+        let scripts_dir = PathBuf::from(format!("{}{}", base_dir_raw, NODE_SCRIPTS_DIR));
+        self.filesystem.create_dir(&base_dir).await?;
+        self.filesystem.create_dir(&config_dir).await?;
+        self.filesystem.create_dir(&data_dir).await?;
+        self.filesystem.create_dir(&scripts_dir).await?;
```

Comment:
Do we need to await each one to happen? Can they all run in parallel and when all are done to fulfill the await? (something like `Promise.all` in JS)

Reply:
Good catch ! Here we can use [join!](https://rust-lang.github.io/async-book/06_multiple_futures/02_join.html) but only for the 3 lasts call because the base_dir needs to be create all 3.